src/parser/Parser.ts

```diff
@@ -98,65 +100,58 @@ export class Parser {
      */
     public ast: Body;
 
-    //TODO remove this once we have verified all of the tests.
-    public statements: Statement[];
-
-    /**
-     * The list of diagnostics found during the parse process
-     */
-    public diagnostics: Diagnostic[];
-
-    /**
-     * The depth of the calls to function declarations. Helps some checks know if they are at the root or not.
-     */
-    private namespaceAndFunctionDepth: number;
-
-    /**
-     * The options used to parse the file
-     */
-    public options: ParseOptions;
-
-    private globalTerminators = [] as TokenKind[][];
-
-    /**
-     * All class statements defined in this file
-     */
-    public classStatements = [] as ClassStatement[];
+    public get statements() {
+        return this.ast.statements;
+    }
 
     /**
-     * All namespace statements defined in this file
+     * References for significant statements/expressions in the parser.
+     * These are initially extracted during parse-time to improve performance, but will also be dynamically regenerated if need be.
+     *
+     * If a plugin modifies the AST, then the plugin should call Parser#clearReferences() to force this list to refresh
      */
-    public namespaceStatements = [] as NamespaceStatement[];
+    public get references() {
+        //build the references object if it's missing.
+        if (!this._references) {
+            this._references = this.findReferences();
+        }
+        return this._references;
+    }
 
-    /**
-     * All function statements defined in this file
-     */
-    public functionStatements = [] as FunctionStatement[];
+    private _references: References = {
```

Comment:
Looks like there is no reason to have all the optional chaining and `?? []`. Optional guard generates quite a lot of overhead.

Reply:
Agreed. I have removed all of the optional chaining stuff.